### PR TITLE
lib: sandbox: refactor to memory store

### PR DESCRIPTION
**Pull request by runcom**
Signed-off-by: Antonio Murdaca <runcom@redhat.com>

**- What I did**

Moved the sandboxes state from a raw map to a Sandbox storer, which is the same as we use in for container objects.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: lib/sandbox/memory_store.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	return &memoryStore{
+		s: make(map[string]*Sandbox),
+	}
+}
+
+// Add appends a new sandbox to the memory store.
+// It overrides the id if it existed before.
+func (c *memoryStore) Add(id string, cont *Sandbox) {
+	c.Lock()
+	c.s[id] = cont
```
Comment (TomSweeneyRedHat) [inline lib/sandbox/memory_store.go]: Golang question.  Is it better to do a defer unlock here (and a couple spots below) just on the extremely small chance that there might be an error before you get to the unlock call?

File: lib/sandbox/memory_store.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+	c.Unlock()
+}
+
+// Get returns a sandbox from the store by id.
+func (c *memoryStore) Get(id string) *Sandbox {
+	c.RLock()
+	res := c.s[id]
```
Comment (TomSweeneyRedHat) [inline lib/sandbox/memory_store.go]: I think you could do a defer RUnlock here and then just return c.s[id]

File: lib/sandbox/store.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+	// Get returns a sandbox from the store by the identifier it was stored with.
+	Get(string) *Sandbox
+	// Delete removes a sandbox from the store by the identifier it was stored with.
+	Delete(string)
+	// List returns a list of sandboxes from the store.
+	List() []*Sandbox
+	// Size returns the number of sandboxes in the store.
+	Size() int
+	// First returns the first sandbox found in the store by a given filter.
+	First(StoreFilter) *Sandbox
```
Comment (TomSweeneyRedHat) [inline lib/sandbox/store.go]: Any use for  a Last() function to get the newest/last sandbox added?

File: lib/sandbox/memory_store.go (hunk 1 above)
Comment (runcom) [reply lib/sandbox/memory_store.go]: nope, it's better to do this as defer as runtime cost, this is a simple case and there cannot be any error till the call to Unlock

File: lib/sandbox/memory_store.go (hunk 2 above)
Comment (runcom) [reply lib/sandbox/memory_store.go]: nope, runtime cost, see comment above

File: lib/sandbox/store.go (hunk 3 above)
Comment (runcom) [reply lib/sandbox/store.go]: Nope, no need for a Last

File: lib/container_server.go [hunk 4]
```diff
@@ -706,60 +706,56 @@ func (c *ContainerServer) ListContainers(filters ...func(*oci.Container) bool) (
...
 }
 
 // RemoveSandbox removes a sandbox from the state store
 func (c *ContainerServer) RemoveSandbox(id string) {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
-	processLabel := c.state.sandboxes[id].ProcessLabel()
-	delete(c.state.sandboxes, id)
+	sb := c.state.sandboxes.Get(id)
+	processLabel := sb.ProcessLabel()
```
Comment (rhatdan) [inline lib/container_server.go]: If sandbox id does not exists, won't this cause a crash?

File: lib/sandbox/history.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+type History []*Sandbox
+
+// Len returns the number of sandboxes in the history.
+func (history *History) Len() int {
+	return len(*history)
+}
+
+// Less compares two sandboxes and returns true if the second one
+// was created before the first one.
+func (history *History) Less(i, j int) bool {
```
Comment (rhatdan) [inline lib/sandbox/history.go]: Why not call it Before?

Comment (runcom) [reply lib/sandbox/history.go]: It's the standard sort interface, I won't work if we call it Before :(

File: lib/container_server.go (hunk 4 above)
Comment (runcom) [reply lib/container_server.go]: Mm yeah we could add a check not to crash, if sb is nil we are in huge issues also

File: lib/sandbox/history.go (hunk 5 above)
Comment (rhatdan) [reply lib/sandbox/history.go]: Ok.

File: lib/sandbox/memory_store.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply lib/sandbox/memory_store.go]: OK thanks for the background info, much appreciated.

File: lib/sandbox/memory_store.go (hunk 2 above)
Comment (TomSweeneyRedHat) [reply lib/sandbox/memory_store.go]: 👍

**Pull request merged**